File: src/0301-0400/383 - Ransom Note/ransom_note.go
package main

func canConstruct(ransomNote string, magazine string) bool {
	if len(ransomNote) > len(magazine) {
		return false
	}
	// make a map of the magazine
	mag := make(map[rune]int)

	// count the number of each character in the magazine
	for _, c := range magazine {
		mag[c]++
	}

	// check if the ransom note can be made from the magazine
	for _, c := range ransomNote {
		if mag[c] == 0 {
			return false
		}
		mag[c]--
	}

	// // make a vector of the magazine
	// magVec := make([]int, 26)

	// for _, c := range magazine {
	// 	magVec[c-'a']++
	// }

	// // check if ransomNote can be constructed from magazine
	// for _, c := range ransomNote {
	// 	if magVec[c-'a'] == 0 {
	// 		return false
	// 	}
	// 	magVec[c-'a']--
	// }

	return true
}

File: src/0301-0400/383 - Ransom Note/ransom_note.rs
impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        // Convert the magazine string to a vector of characters
        let mut magazine = magazine.chars().collect::<Vec<char>>();

        // Iterate over the ransom note string
        for c in ransom_note.chars() {
            // If the magazine contains the character, remove it
            // Otherwise, return false
            // position() returns the index of the first element that matches the closure
            // else returns None
            if let Some(i) = magazine.iter().position(|&x| x == c) {
                magazine.remove(i);
            } else {
                return false;
            }
        }

        true
    }
}
